custom/enikshay/user_setup.py

```diff
@@ -226,18 +226,11 @@ def compress_id(serial_id, growth_symbols, lead_symbols, body_symbols, body_digi
     return ''.join(output)
 
 
-def get_last_used_device_number(user):
-    if not user.devices:
-        return None
-    _, index = max((device.last_used, i) for i, device in enumerate(user.devices))
-    return index + 1
-
-
-def set_enikshay_device_id(user):
-    device_number = get_last_used_device_number(user)
-    if device_number and user.user_data.get('id_device_number', None) != device_number:
-        user.user_data['id_device_number'] = device_number
-        user.user_data['id_device_body'] = compress_nikshay_id(device_number, 0)
+def set_enikshay_device_id(user, device_id):
+    # device_id was JUST set, so it must be in there
+    device_number = [device.device_id for device in user.devices].index(device_id) + 1
```

Comment:
does the order of user.devices never change?

Reply:
It doesn't - it's modified here using append: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/users/models.py#L2123-L2123

Reply:
isn't this part still going to be an issue if there's no locking, since editing two user objects at the same time would lead to a doc update conflict?

Reply:
Hey, sorry, I didn't notice this comment until just now.  I now believe that race conditions are not the primary issue (although they are a possibility).  My last two commits address the more probable cause.  I think this change should address the immediate issue, and then medium term we'll switch off of this approach altogether onto something more robust (that doesn't involve saving this to the user), particularly if we keep having issues.